examples/4_totp_storage.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+
+###################################
+# WHAT IS IN THIS EXAMPLE?
+#
+# Keybase has added an encrypted key-value store; see 4_simple_storage.py for
+# more information.
+#
+# This example shows how you can build a simple TOTP bot that makes use of
+# the team encrypted key-value store, which we interact with using KVStoreClient.
+#
+# This example does minimal error handling and is not concurrency safe.
+###################################
+
+import asyncio
+import json
+import logging
+import os
+import sys
+from enum import Enum
+
+import pyotp
+
+from pykeybasebot import Bot, EventType
+
+logging.basicConfig(level=logging.DEBUG)
+
+if "win32" in sys.platform:
+    # Windows specific event-loop policy
+    asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
+
+
+class TotpMsg(Enum):
+    PROVISION = "provision"
+    REPROVISION = "reprovision"
+    REMOVE = "remove"
+    NOW = "now"
+    URI = "uri"
+    LIST = "list"
+
+
+class TotpHandler:
+    """
+    TotpHandler handles commands sent via chat and uses the team key-value store to
+    provide TOTP client functionality.
+
+    TotpHandler listens to chat messages of the form:
+
+    `!totp {provision|reprovision|remove|now|uri} <issuer>` and
+    `!totp list`
+
+    For each provisioned key, the handler stores in the namespace "totp" one
+    row, with the key "<issuer" and the json blob value "{"secret": <16 char base32
+    secret>}".
+
+    This Keybase bot can be used in place of MFA apps like Google Authenticator and Authy
+    for login processes that require two-step verification. This bot uses the
+    pyotp library, so it is also compatible with other MFA apps.
+
+    A Keybase TOTP bot could be useful for logins where you're already using a strong
+    password, but are required to also use TOTP. Or it could be perfect for adding
+    replay attack resistance to the physical keypad lock protecting your team's ice
+    cream freezer.... Note that depending on the login process and how you use this bot
+    with your Keybase team setup, this bot may not be appropriate for threat models
+    that require a physical "second factor" device.
+
+    For more information on TOTP, see https://pyotp.readthedocs.io/en/latest/,
+    https://tools.ietf.org/html/rfc6238.
+    """
+
+    MSG_PREFIX = "!totp"
+    NAMESPACE = "totp"
+
+    def to_json(self, secret):
+        return {"secret": secret}
+
+    async def __provision(self, bot, team, issuer, force: bool = False):
+        secret = pyotp.random_base32()
+        val = json.dumps(self.to_json(secret))
+        if not force:
+            try:
+                # throws exception if 1 is not the latest revision + 1
+                await bot.kvstore.put(team, self.NAMESPACE, issuer, val, revision=1)
+            except Exception as e:
```

Comment:
suggest we inspect the exception. it _should_ have a `code` field with an integer value that can be used.

Reply:
will tackle with the 3rd encrypted entryKey example